### PR TITLE
Add position ids in forward pass to opt model

### PR DESCRIPTION
**Pull request by avishaiElmakies**
# What does this PR do?

This pull request adds position_ids to the forward of ```OPT``` in a similar fashion to ```gemma``` and ```llama```. #32937 

Some models didn't have an argument for position_ids in their forward pass. 

There are two main reasons we would like for all LM models to get positions ids.

1. to have the API be consistent with all models.
2. position_ids are very important if you want to use flash-attention without padding, during training. if i want to be able to pack two or more sentences in the same sequence. I would like to know that the model handles the sentences accordingly and treats each sentence as it's own different sentence. flash-attention code uses position_ids to check if some sequences are packed and runs an appropriate function to make sure there is no cross-example contamination. but without this, the model can't use this feature. the code always checks if position_ids is not None:

https://github.com/huggingface/transformers/blob/v4.44.1/src/transformers/modeling_flash_attention_utils.py#L270

This handles only OPT, so i can start small and get some feedback.

changes:

- changed [OPTLearnedPositionalEmbedding](https://github.com/huggingface/transformers/blob/v4.44.2/src/transformers/models/opt/modeling_opt.py#L63) forward method to get position_ids instead of attention_mask.
- changed all forward passes in the file to get position_ids and pass them forward. 
- update prepare_inputs_for_generation to get position_ids. 
- if position_ids are ```None``` create position_ids based on attention (similar to original version, so it should work the same if position_ids are not given)
- update relevent docs

a few notes:

- this model because of the use of offset in [OPTLearnedPositionalEmbedding](https://github.com/huggingface/transformers/blob/v4.44.2/src/transformers/models/opt/modeling_opt.py#L63) needs to represent it's padding token position as -1. I think this is fine to keep compatibility with the weights and everything. 
- [BioGPT](https://github.com/huggingface/transformers/blob/v4.44.2/src/transformers/models/biogpt/modeling_biogpt.py#L50) seems to copy OPT's positional embedding class. I needed to remove the comment to disable that (since it no longer copies). this is needed for ```make fixup```.

feature-request #32937 

## Before submitting
- [] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [] Did you write any new necessary tests?

@ArthurZucker would love some feedback.

### CODE REVIEW

File: src/transformers/models/opt/modeling_opt.py [hunk 1]
```diff
@@ -60,6 +60,14 @@
 _SEQ_CLASS_EXPECTED_OUTPUT = "'LABEL_0'"
 
 
+def _attention_to_position_ids(attention_mask: torch.Tensor) -> torch.Tensor:
+    """
+    Converts an attention mask to position ids.
+    """
+
+    return (torch.cumsum(attention_mask, dim=1) * attention_mask).long() - 1
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:68]: let's avoid such one liners please

File: src/transformers/models/biogpt/modeling_biogpt.py [hunk 2]
```diff
@@ -46,7 +46,6 @@
 _CONFIG_FOR_DOC = "BioGptConfig"
 
 
-# Copied from transformers.models.opt.modeling_opt.OPTLearnedPositionalEmbedding with OPT->BioGpt
```
Comment (ArthurZucker) [inline src/transformers/models/biogpt/modeling_biogpt.py:49]: let's just add a TODO here or also update that model

File: src/transformers/models/opt/modeling_opt.py [hunk 3]
```diff
@@ -71,17 +79,10 @@ def __init__(self, num_embeddings: int, embedding_dim: int):
         self.offset = 2
         super().__init__(num_embeddings + self.offset, embedding_dim)
 
-    def forward(self, attention_mask: torch.LongTensor, past_key_values_length: int = 0):
+    def forward(self, position_ids: torch.LongTensor):
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:82]: that is kind of a breaking change for this module 😓

File: src/transformers/models/opt/modeling_opt.py [hunk 4]
```diff
@@ -1060,6 +1096,14 @@ def prepare_inputs_for_generation(
 
             input_ids = input_ids[:, remove_prefix_length:]
 
+        if attention_mask is not None and position_ids is None:
+            position_ids = _attention_to_position_ids(attention_mask)
+            if past_key_values:
+                position_ids = position_ids[:, -input_ids.shape[1] :]
+
+                # This `clone` call is needed to avoid recapturing cuda graphs with `torch.compile`'s  `mode="reduce-overhead`, as otherwise the input `position_ids` would have various stride during the decoding. Here, simply using `.contiguous()` is not sufficient as in the batch size = 1 case, `position_ids` is already contiguous but with varying stride which retriggers a capture.
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:1104]: comment is way to long let's split it on more lines!

**Review by ArthurZucker — COMMENTED** [review]
Hey! in general the trick is that adding a new argument needs to be done at the end of the forward pass otherwise you are breaking the model for people who directly call the model

File: src/transformers/models/opt/modeling_opt.py (hunk 1 above)
Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:68]: OK, I will change that.

File: src/transformers/models/opt/modeling_opt.py (hunk 3 above)
Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:82]: why is this a problem? the weights are the same so loading should work. and this module should not be used by outside code so it is not supposed to break anything.

File: src/transformers/models/biogpt/modeling_biogpt.py (hunk 2 above)
Comment (avishaiElmakies) [reply src/transformers/models/biogpt/modeling_biogpt.py:49]: should I return the comment as well? it causes a fail when using make fixup.
I tried to look to update BioGPT, but it seems to be a lot of code from different models, so i didn't know if i should have touched it. I can work on it next.

File: src/transformers/models/opt/modeling_opt.py (hunk 4 above)
Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:1104]: K

**Comment by avishaiElmakies** [comment]
I tried to keep the API as similar as possible to ```llama``` and ```gamma```. they both put it third. isn't it kinda of a problem with AutoModel to have them at different placements?  it means that you can't load using AutoModel and use the model without keyword arguments?

File: src/transformers/models/biogpt/modeling_biogpt.py (hunk 2 above)
Comment (ArthurZucker) [inline src/transformers/models/biogpt/modeling_biogpt.py:49]:
```suggestion
# copied from transformers.models.opt.modeling_opt.OPTLearnedPositionalEmbedding with OPT->BioGpt
# TODO @ArthurZucker bring copied from back

File: src/transformers/models/opt/modeling_opt.py [hunk 5]
```diff
@@ -60,6 +60,15 @@
 _SEQ_CLASS_EXPECTED_OUTPUT = "'LABEL_0'"
 
 
+def _attention_to_position_ids(attention_mask: torch.Tensor) -> torch.Tensor:
+    """
+    Converts an attention mask to position ids.
+    """
+
+    position_ids = torch.cumsum(attention_mask, dim=1)
+    return (position_ids * attention_mask - 1).long()
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:69]: Humm let's avoid such one liners

File: src/transformers/models/opt/modeling_opt.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/opt/modeling_opt.py:83]: Yeah but it has caused issue in the past 😉

File: src/transformers/models/opt/modeling_opt.py [hunk 6]
```diff
@@ -123,6 +125,9 @@ def forward(
         self,
         hidden_states: torch.Tensor,
         key_value_states: Optional[torch.Tensor] = None,
+        position_ids: Optional[
+            torch.Tensor
+        ] = None,  # isn't needed in normal attention, but needed in flash attention so to keep the signature same
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:130]: should be added at the end of the list of args

File: src/transformers/models/opt/modeling_opt.py [hunk 7]
```diff
@@ -260,6 +265,7 @@ def forward(
         self,
         hidden_states: torch.Tensor,
         key_value_states: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.Tensor] = None,
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:268]: same hjere

**Review by ArthurZucker — COMMENTED** [review]
cc @gante WDYT about this? In general IMO we should just run basic position ids init. Tho taking padding into account should be "alright", it's already done in generate, this would help for forward and training. 

Just need to be careful as we also want to support packing

File: src/transformers/models/opt/modeling_opt.py (hunk 6 above)
Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:130]: OK, will change those. 

about the forward call for the embedding layer. I think it has to take position ids as an argument. otherwise it will not work with packed sentences.

File: src/transformers/models/opt/modeling_opt.py (hunk 7 above)
Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:268]: k

**Comment by avishaiElmakies** [comment]
@ArthurZucker I am thinking that maybe the best solution for the embedding layer is to add position_ids as an arg to the forward pass with default None. this is probably backward compatible, but will still help with packed sentences. the problem is probably that the code will not be very nice

**Review by gante — COMMENTED** [review]
@ArthurZucker I'm pro `position_ids` as it standardizes OPT wrt other models 🙌 

@avishaiElmakies Thank you for adding the fix 🤗 Have a look at unresolved comments (you'd be surprised with how easy it is to break code for other external libraries, [hyrum's law](https://www.hyrumslaw.com/) definitely applies to `transformers`)

File: src/transformers/models/opt/modeling_opt.py (hunk 3 above)
Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:83]: OK, how do you think I should do it? The module should get position_ids to work with packed sentences. Should I add position_ids as the last argument with None as default?

**Comment by avishaiElmakies** [comment]
@gante, thanks! Happy to contribute.

I would love some guidance on the last two comments left, what should I do with the position_ids in the embedding module. In my opion it should be able to get position_ids to work with packed sentences. Maybe a last argument with default None and a check?

And about the one liners, would love some guidance on that

**Comment by avishaiElmakies** [comment]
@ArthurZucker would love some guidance here so i can finish and move on to other models.

File: src/transformers/models/opt/modeling_opt.py (hunk 5 above)
Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:69]: OK, now i understand. I will change that! thanks!

**Comment by avishaiElmakies** [comment]
@ArthurZucker changed what you said and refactored the embedding class to be backward compatible. would love some feedback

File: src/transformers/models/biogpt/modeling_biogpt.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/biogpt/modeling_biogpt.py:50]: yes !

File: src/transformers/models/opt/modeling_opt.py [hunk 8]
```diff
@@ -724,7 +745,13 @@ def forward(
                 attention_mask, input_shape, inputs_embeds, past_key_values_length
             )
 
-        pos_embeds = self.embed_positions(attention_mask, past_key_values_length)
+        if position_ids is None:
+            position_ids = torch.cumsum(attention_mask, dim=1)
+            position_ids = (position_ids * attention_mask - 1).long()
+            # cut positions if `past_key_values_length` is > 0
+            position_ids = position_ids[:, past_key_values_length:]
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:752]: only thing against this is that it's not compile compatible (dynamic slicing on input dependent) but okay!

File: src/transformers/models/opt/modeling_opt.py [hunk 9]
```diff
@@ -1061,6 +1099,18 @@ def prepare_inputs_for_generation(
...
+            position_ids = torch.cumsum(attention_mask, dim=1)
+            position_ids = (position_ids * attention_mask - 1).long()
+            if past_key_values:
+                position_ids = position_ids[:, -input_ids.shape[1] :]
+
+                # This `clone` call is needed to avoid recapturing cuda graphs with `torch.compile`'s  `mode="reduce-overhead`,
+                # as otherwise the input `position_ids` would have various stride during the decoding.
+                # Here, simply using `.contiguous()` is not sufficient as in the batch size = 1 case,
+                # position_ids` is already contiguous but with varying stride which retriggers a capture.
+                position_ids = position_ids.clone(memory_format=torch.contiguous_format)
```
Comment (ArthurZucker) [inline src/transformers/models/opt/modeling_opt.py:1112]: @gante can confirm but `generate` prepares them no? SHould not be needed AFAIK!

**Review by ArthurZucker — APPROVED** [review]
Feel free to merge @gante if it's alright with you! 🤗 and thanks for your contribution!

File: src/transformers/models/opt/modeling_opt.py (hunk 8 above)
Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:752]: is there a way to make it compile compatible?

Comment (ArthurZucker) [reply src/transformers/models/opt/modeling_opt.py:752]: creating a tensor with `torch.arange(past_key_values_length, full_length)` but we would need to test!

Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:752]: Ok, do you want me to try? and how do you test if it is compile compatiable?

Comment (ArthurZucker) [reply src/transformers/models/opt/modeling_opt.py:752]: We have a slow test for this: https://github.com/huggingface/transformers/blob/fe496262c119e4366b2f319567bb561cab445dc1/tests/test_modeling_common.py#L4986 to make sure it runs!

Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:752]: how can i make the ```torch.arange(past_key_values_length, full_length)``` work with multiplying attention_mask (similar to ```position_ids * attention_mask```. it will need dynamic slicing as well?
does dynamic slicing compile if it does not under ```if position_ids is None```?

Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:752]: @ArthurZucker after looking at the test, I think it will not work for opt. the test uses static cache and opt doesn't support static cache. So making compile work for OTP will need further work

Comment (ArthurZucker) [reply src/transformers/models/opt/modeling_opt.py:752]: Ahhh right, I thought #32617 had touched OPT, but not yet! 
Then no worries

File: src/transformers/models/opt/modeling_opt.py (hunk 9 above)
Comment (ArthurZucker) [reply src/transformers/models/opt/modeling_opt.py:1112]: IMO let's remove it from prepare input for generation at least to de-duplicate

File: src/transformers/models/opt/modeling_opt.py (hunk 8 above)
Comment (avishaiElmakies) [reply src/transformers/models/opt/modeling_opt.py:752]: OK
